fix spammy library deinit errors when user passes an invalid TORCH_LOGS argument

bdhirsh opened the pull request:
fixes https://github.com/pytorch/pytorch/issues/151055. Thanks @desertfire for the patch that fixed this.

I was a bit careful about the test - I wanted to make sure the test accurately ensures that we don't regress and our error message is not spammy when users enter an invalid `TORCH_LOGS=....` argument. But I tried to avoid using expecttests, since people  occasionally add new logging artifacts and I didn't want to add to much churn by forcing this to fail CI.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151678



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh committed:
fix spammy library deinit errors when user passes an invalid TORCH_LOGS argument

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

zou3519 reviewed (approved):
looks fine, but what was the problem and how did this fix it?

zou3519 commented on torch/library.py:
```diff
@@ -452,7 +452,7 @@ def _del_library(
     registration_handles,
 ):
-    import torch.fx
+    from torch.fx.operator_schemas import _SCHEMA_TO_SIGNATURE_CACHE
```
Ugh i think the problem is that the finalizer isn't allowed to import anything, because it might be run when python is exiting.

If that's the case _del_library should take _SCHEMA_TO_SIGNATURE_CACHE as an input and then mutate it.

zou3519 replied:
Another reason why it should take that variable is that it might be gone by the time the finalizer runs too

bdhirsh replied:
this is a good catch. updating

zou3519 reviewed (changes requested):
see comment

bdhirsh committed:
Update on "fix spammy library deinit errors when user passes an invalid TORCH_LOGS argument"

fixes https://github.com/pytorch/pytorch/issues/151055. Thanks desertfire for the patch that fixed this.

I was a bit careful about the test - I wanted to make sure the test accurately ensures that we don't regress and our error message is not spammy when users enter an invalid `TORCH_LOGS=....` argument. But I tried to avoid using expecttests, since people  occasionally add new logging artifacts and I didn't want to add to much churn by forcing this to fail CI.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "fix spammy library deinit errors when user passes an invalid TORCH_LOGS argument"

fixes https://github.com/pytorch/pytorch/issues/151055. Thanks desertfire for the patch that fixed this.

I was a bit careful about the test - I wanted to make sure the test accurately ensures that we don't regress and our error message is not spammy when users enter an invalid `TORCH_LOGS=....` argument. But I tried to avoid using expecttests, since people  occasionally add new logging artifacts and I didn't want to add to much churn by forcing this to fail CI.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
